kaolin/datasets/base.py

```diff
@@ -0,0 +1,154 @@
+        """
+        self.initialize(*args, **kwargs)
+        if preprocessing_transform is not None:
+            desc = 'applying preprocessing'
+            if preprocessing_params is None:
+                preprocessing_params = {}
+            assert preprocessing_params.get('cache_dir') is not None
+            self.cache_convert = helpers.Cache(
+                preprocessing_transform, preprocessing_params['cache_dir'],
+                cache_key=helpers._get_hash(repr(preprocessing_transform)))
+            if preprocessing_params.get('use_cuda') is None:
+                preprocessing_params['use_cuda'] = False
+            num_workers = preprocessing_params.get('num_workers')
+            if num_workers == 0:
+                with torch.no_grad():
+                    for idx in tqdm(range(len(self)), desc=desc, disable=no_progress):
+                        name = self._get_attributes(idx)['name']
+                        if name not in self.cache_convert.cached_ids:
+                            data = self._get_data(idx)
+                            self.cache_convert(name, *data)
+            else:
+                p = Pool(num_workers)
+                iterator = p.imap_unordered(
+                    _preprocess_task,
+                    [(idx, self._get_data, self._get_attributes, self.cache_convert)
+                     for idx in range(len(self))])
+                for i in tqdm(range(len(self)), desc=desc, disable=no_progress):
+                    next(iterator)
+        else:
+            self.cache_convert = None
+        self.transform = transform
+
+    def __getitem__(self, index):
+        """Returns the item at index idx. """
+        attributes = self._get_attributes(index)
+        data = (self.cache_convert(attributes['name']) if self.cache_convert is not None else
+                self._get_data(index))
+        if self.transform is not None:
+            data = self.transform(data)
+        return {'data': data, 'attributes': attributes}
+
+    @abstractmethod
+    def initialize(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def _get_attributes(self, index):
+        pass
+
+    @abstractmethod
+    def _get_data(self, index):
+        pass
+
+    @abstractmethod
+    def __len__(self):
+        pass
+
+class CombinationDataset(KaolinDataset):
+    """Dataset combining multiple list of datasets outputs as inputs
+    the output of _get_attributes will be a tuple of all the _get_attributes of the dataset list
+    the output of _get_data wiil be a tuple of all the 'data' of the __getitem__ of the dataset list
+
+    Args:
+        datasets: list or tuple of KaolinDataset
+    """
+    def _initialize(self, datasets):
+        self.len = len(datasets[0])
+        for i, d in enumerate(datasets):
+            assert len(d) == self.len, \
+                "dataset {} have different length ({})than the first dataset ({})".format(i, len(d), self.len)
```

Comment:
Can change the error message to be more descriptive
```
f"All datasets must have the same length. Invalid length at index {i} (expected: {self.len}, got: {len(d)})"
```
Note: used fstring only for convenience

Reply:
That's better, done